### PR TITLE
[ARITH]add simplify rule for div

### PR DESCRIPTION
**Pull request by xqdan**
Thanks for contributing to TVM!   Please refer to guideline https://tvm.apache.org/docs/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

@tqchen @yzhliu please review

I add some proof here,  just consider (x+c2)//c3 - (x+c4)//c3, this is the rule we need,

one condition at SubNode we have is c2<c3 and c4<c3

1) when x+c2<=c3 and x+c4<=c3, it's safe to combine them, so x<min(c3-c2, c3-c4) 
2) when x+c2>c3 and x+c4>c3,  and (x+c2-x-c4)%c3 == 0, it's safe to combine them. so x>max(c3-c2, c3-c4) and (c2-c4)%c3==0. since  c2<c3 and c4<c3, so x>max(c3-c2, c3-c4) is enough.

@zhiics @tqchen please check

### CODE REVIEW

**Comment by tqchen** [comment]
cc @yongfeng-nv @yzhliu @zhiics @wweic @junrushao1994 please help to take a look

File: src/arith/rewrite_simplify.cc [hunk 1]
```diff
@@ -353,6 +353,19 @@ PrimExpr RewriteSimplifier::Impl::VisitExpr_(const SubNode* op) {
         truncdiv(x + c1, c3) - truncdiv(x, c3), truncdiv(truncmod(x, c3) + c1, c3),
         CanProveGreaterEqual(x.Eval(), 0) && c1.Eval()->value >= 0 && c3.Eval()->value > 0);
 
+    TVM_TRY_REWRITE_IF(truncdiv(x * c1 + c2, c3) - truncdiv(x * c1 + c4, c3), truncdiv(c2 - c4, c3),
+                       c1.Eval()->value >= 0 && c2.Eval()->value > 0 && c3.Eval()->value > 0 &&
+                           c4.Eval()->value >= 0 && CanProveGreaterEqual(x.Eval(), 0));
```
Comment (zhiics) [inline src/arith/rewrite_simplify.cc:358]: hmm,  I am wondering if this always holds.  

say `x = 0, c1 = 1, c2 = 3, c3 = 3, c4 = 1`
`truncdiv(x * c1 + c2, c3) - truncdiv(x * c1 + c4, c3) = = truncdiv(3, 3) - truncdiv(1, 3)`
`truncdiv(c2 - c4, c3) = truncdiv(2, 3)`

Comment (xqdan) [reply src/arith/rewrite_simplify.cc:358]: Yes, need to add condition: c2-c4>=c3

**Comment by tqchen** [comment]
cc @zhiics @yongfeng-nv @merrymercy @jroesch please help to take a look

File: src/arith/rewrite_simplify.cc [hunk 2]
```diff
@@ -353,6 +353,25 @@ PrimExpr RewriteSimplifier::Impl::VisitExpr_(const SubNode* op) {
         truncdiv(x + c1, c3) - truncdiv(x, c3), truncdiv(truncmod(x, c3) + c1, c3),
         CanProveGreaterEqual(x.Eval(), 0) && c1.Eval()->value >= 0 && c3.Eval()->value > 0);
 
+    TVM_TRY_REWRITE_IF(truncdiv(x * c1 + c2, c3) - truncdiv(x * c1 + c4, c3), truncdiv(c2 - c4, c3),
+                       c1.Eval()->value >= 0 && c2.Eval()->value > 0 && c3.Eval()->value > 0 &&
+                           c4.Eval()->value >= 0 && CanProveGreaterEqual(x.Eval(), 0) &&
+                           std::abs(c4.Eval()->value - c2.Eval()->value) >= c3.Eval()->value);
```
Comment (zhiics) [inline src/arith/rewrite_simplify.cc:359]: hmm, is this always true again?

say `c1 = 0, c2 = 6, c3 = 3, c4 = 2`, `c4 - c2 = 4 > 3`
`truncdiv(x * c1 + c2, c3) - truncdiv(x * c1 + c4, c3) = truncdiv(6, 3) - truncdiv(2, 3)`
`truncdiv(c2 - c4, c3) = truncdiv(4, 3)`

Comment (xqdan) [reply src/arith/rewrite_simplify.cc:359]: I add your case in unitest. truncdiv(6, 3)  will be simplified to 2 first, so what we catch here is '2-tdiv(2,3)'.

Comment (xqdan) [reply src/arith/rewrite_simplify.cc:359]: truncdiv(6, y) - truncdiv(2, y) if y=3 will break the rule, remove y related rule for now.

**Comment by zhiics** [comment]
LGTM now. @yongfeng-nv @merrymercy @junrushao1994 any of you wants to take a quick look?

File: src/arith/rewrite_simplify.cc [hunk 3]
```diff
@@ -353,6 +353,15 @@ PrimExpr RewriteSimplifier::Impl::VisitExpr_(const SubNode* op) {
         truncdiv(x + c1, c3) - truncdiv(x, c3), truncdiv(truncmod(x, c3) + c1, c3),
         CanProveGreaterEqual(x.Eval(), 0) && c1.Eval()->value >= 0 && c3.Eval()->value > 0);
 
+    TVM_TRY_REWRITE_IF(truncdiv(x * c1 + c2, c3) - truncdiv(x * c1 + c4, c3), truncdiv(c2 - c4, c3),
+                       c1.Eval()->value >= 0 && c2.Eval()->value > 0 && c3.Eval()->value > 0 &&
```
Comment (tqchen) [inline src/arith/rewrite_simplify.cc:357]: Would be great if we can write down a proof for the rule. or a tabular validation. This rule is quite similar to the rule above(which translates into an additional truncmod), but this rule does not.

**Comment by tqchen** [comment]
Arith simplification are key to the correctness so it would be great if we can be extra careful here. It would be great if we can write down a proof for the rule, or at least try to enumerate two tabular values to see if all of the possibilities within a range holds.

File: src/arith/rewrite_simplify.cc (hunk 3 above)
Comment (xqdan) [reply src/arith/rewrite_simplify.cc:357]: ok, I add some proof above, please check

**Comment by tqchen** [comment]
cc @spectrometerHBH @Hzfengsy would be great if you can also help  to check the proof

**Comment by tqchen** [comment]
@xqdan sorry for the delayed reply. I take a closer look, and unfortunately the new rules may not be correct.

```
    TVM_TRY_REWRITE_IF(
        truncdiv(x + c1, c3) - truncdiv(x, c3), truncdiv(truncmod(x, c3) + c1, c3),
        CanProveGreaterEqual(x.Eval(), 0) && c1.Eval()->value >= 0 && c3.Eval()->value > 0);
```

In particular, the new rule is closely related to the above rule(which is on top of the current rule. Pay attention to the special case of `truncmod(x, c3)` in the above term.

In particular, let us say `c3=3, c1=1, x = 11`, then
```
truncdiv(11 + 1, 3) - truncdiv(11, 3) = 4 - 3 = 1 = truncdiv(truncmod(11, 3)+1, 3)
```

Let `x =y * c4 + c5` and replace `y` by x, we can derive a counter-example for the listed rule.  To re-use the above rule a bit, a likely generalization should be.

It is not too hard to prove the above rule(by decomposing x into z * c3 + r).

To get to a similar rule that applies to your case, we could replace x by x + c4, then get to

```
    TVM_TRY_REWRITE_IF(
        truncdiv(x + c4 + c1, c3) - truncdiv(x + c4, c3), truncdiv(truncmod(x + c4, c3) + c1, c3),
        CanProveGreaterEqual((x + c4).Eval(), 0) && c1.Eval()->value >= 0 && c3.Eval()->value > 0);
```

=>

```
    TVM_TRY_REWRITE_IF(
        truncdiv(x + c5, c3) - truncdiv(x + c4, c3), truncdiv(truncmod(x + c4, c3) + (c5-c4), c3),
        CanProveGreaterEqual((x + c4).Eval(), 0) && (c5 - c4).Eval()->value >= 0 && c3.Eval()->value > 0);
```

**Comment by xqdan** [comment]
> @xqdan sorry for the delayed reply. I take a closer look, and unfortunately the new rules may not be correct.
> 
> ```
>     TVM_TRY_REWRITE_IF(
>         truncdiv(x + c1, c3) - truncdiv(x, c3), truncdiv(truncmod(x, c3) + c1, c3),
>         CanProveGreaterEqual(x.Eval(), 0) && c1.Eval()->value >= 0 && c3.Eval()->value > 0);
> ```
> 
> In particular, the new rule is closely related to the above rule(which is on top of the current rule. Pay attention to the special case of `truncmod(x, c3)` in the above term.
> 
> In particular, let us say `c3=3, c1=1, x = 11`, then
> 
> ```
> truncdiv(11 + 1, 3) - truncdiv(11, 3) = 4 - 3 = 1 = truncdiv(truncmod(11, 3)+1, 3)
> ```
> 
> Let `x =y * c4 + c5` and replace `y` by x, we can derive a counter-example for the listed rule. To re-use the above rule a bit, a likely generalization should be.
> 
> It is not too hard to prove the above rule(by decomposing x into z * c3 + r).
> 
> To get to a similar rule that applies to your case, we could replace x by x + c4, then get to
> 
> ```
>     TVM_TRY_REWRITE_IF(
>         truncdiv(x + c4 + c1, c3) - truncdiv(x + c4, c3), truncdiv(truncmod(x + c4, c3) + c1, c3),
>         CanProveGreaterEqual((x + c4).Eval(), 0) && c1.Eval()->value >= 0 && c3.Eval()->value > 0);
> ```
> 
> =>
> 
> ```
>     TVM_TRY_REWRITE_IF(
>         truncdiv(x + c5, c3) - truncdiv(x + c4, c3), truncdiv(truncmod(x + c4, c3) + (c5-c4), c3),
>         CanProveGreaterEqual((x + c4).Eval(), 0) && (c5 - c4).Eval()->value >= 0 && c3.Eval()->value > 0);
> ```

@tqchen thanks for reply. 
 I get your point, but what we seek is to eliminate 'x' here,
 The new rule 2 above (not in the code) can exclude the case you post, since (11+1 - 11) %3 == 0 is not satisfied.

**Comment by tqchen** [comment]
It would be great to elaborate your conditions in 1, and 2 a bit and write down a new rule :) I can double check a bit later

**Pull request closed without merging**